File: shortest_sub_segment/stuff/shortest_sub_segment.py
#!/usr/bin/env python
# -*- coding: utf-8 -*- 

import re
import binascii
import string

__author__ = "Alonso Vidales"
__email__ = "alonso.vidales@tras2.es"
__date__ = "2013-11-11"

class ShortestSubSegment:
    """
    This is a solution based on hashing, I obtain a integer value for each word,
    and check all the possible subsets, is the sum of all the integer values is
    the same than the sum for all the words, is high probable that this sub set
    contain the given words.
    This kind of comparation is really fast, and have a complexity of O(n)
    where n are the number of words.
    """
    __debug = False

    def resolve(self):
        # Get a list with only words from the words list
        filtered = []
        word_positions = []
        last_word = 0
        for word_pos in range(len(self.__textWords)):
            if self.__textWords[word_pos].lower() in self.__words:
                word = binascii.crc32(self.__textWords[word_pos].lower())
                if last_word != word:
                    last_word = word
                    filtered.append(word)
                    word_positions.append(word_pos)

        if self.__debug:
            print "Filtered: %s" % filtered
            print "Word Pos: %s" % word_positions

        smallest = len(self.__textWords)
        sub_set = False
        # Check the sum of all the words of the sub sets
        for subset_pos in range(len(filtered) - len(self.__words) + 1):
            crc_subset_words = set(filtered[subset_pos:subset_pos + len(self.__words)])

            extra_words = 0
            while len(crc_subset_words) != len(self.__words) and subset_pos + len(self.__words) + extra_words < len(filtered):
                if self.__debug:
                    print "Extra %d - %s" % (subset_pos + len(self.__words) + extra_words, filtered[subset_pos + len(self.__words) + extra_words])
                crc_subset_words.add(filtered[subset_pos + len(self.__words) + extra_words])
                extra_words += 1

            if self.__debug:
                print "Ended: %s" % crc_subset_words

            if sum(crc_subset_words) == self.__wordsWeight:
                first = word_positions[subset_pos]
                last = word_positions[subset_pos + len(self.__words) + extra_words - 1]
                lenght = last - first + 1
                if lenght < smallest:
                    smallest = lenght
                    if self.__debug:
                        print "First: %d - %d - Diff: %d" % (first, last, lenght)

                    sub_set = self.__textWords[first:first + lenght]

                    # Well is pretty sure that we have found the subset, but is
                    # better check it words by word in order to avoid false positives
                    found = True
                    for wordPos in range(len(sub_set)):
                        if sub_set[wordPos].lower() not in self.__words:
                            found = False

                    if found and len(sub_set) == len(self.__words):
                        return ' '.join(sub_set)

        if not sub_set:
            return "NO SUBSEGMENT FOUND"

        return ' '.join(sub_set)

    def __init__(self, text, words):
        self.__textWords = re.sub('[^A-Za-z ]+', '', text).split()
        self.__textWordsCrc = map(binascii.crc32, map(string.lower,self.__textWords))
        self.__words = set(map(string.lower, words))
        self.__wordsWeight = sum(map(binascii.crc32, map(string.lower, words)))

        if self.__debug:
            print "Text Words: %s" % self.__textWords
            print "Text Words Crc: %s" % self.__textWordsCrc
            print "Words: %s" % self.__words_crc
            print "Words Weight: %s" % self.__wordsWeight

if __name__ == "__main__":
    text = raw_input()
    words = [raw_input() for words in range(int(raw_input()))]

    print ShortestSubSegment(text, words).resolve()

File: shortest_sub_segment/stuff/shortest_sub_segment.go
package main

import "fmt"
import "bufio"
import "os"
import "strconv"
import "strings"
import "regexp"
import "hash/crc32"

/**
 * @author Alonso Vidales <alonso.vidales@tras2.es>
 * @date 2013-11-12
 *
 * @description: This solution is based in convert the words in int values using
 *	crc32 and try to find the smallest subset that sum the same than the given
 *	words, doing this, we have a hight probability of find the correct sub set
 */

/**
 * Returns two arrays that includes only the words of the text ho matches with
 * the given words, one of the arrays is a list of the crc32 values of the words
 * and the other one the positions in the original string of this words
 */
func get_words_crc_pos(text []string, available_words map[uint64]bool) (text_word_pos []int, text_word_crc []uint64) {
    // Divide the original text in words, and assign a value to each word to be find
    for pos, word := range text {
	crc32 := uint64(crc32.Checksum([]byte(strings.ToLower(word)), crc32.IEEETable))
	_, found := available_words[crc32]
	if found {
	    text_word_crc = append(text_word_crc, crc32)
	    text_word_pos = append(text_word_pos, pos)
	}
    }

    return text_word_pos, text_word_crc
}

func Sum(a *[]uint64) (sum uint64) {
    for _, v := range *a {
	sum += v
    }
    return
}

func main() {
    bio := bufio.NewReader(os.Stdin)

    text, _ := bio.ReadString('\n')
    text = text[:len(text) - 1]
    words, _ := bio.ReadString('\n')
    words_int, _ := strconv.ParseInt(words[:len(words) - 1], 10, 0)

    var words_total_weight uint64
    words_total_weight = 0

    // Build a set with the given words in order to know if a word is included
    // or not with O(1)
    available_words := make(map[uint64]bool)
    for i:= 0; i < int(words_int); i++ {
	word, _ := bio.ReadString('\n')
	if word[len(word) - 1] == '\n' {
	    word = strings.ToLower(word[:len(word) - 1])
	}
	crc32 := uint64(crc32.Checksum([]byte(word), crc32.IEEETable))
	available_words[crc32] = true
	words_total_weight += crc32
	//fmt.Println("Words weight:", words_total_weight)
    }

    // Filter the string to get only the [a-z][A-Z] chars
    re, _ := regexp.Compile(`[^A-Za-z ]+`)
    text = string(re.ReplaceAll([]byte(text), []byte("")))
    //fmt.Println("After:", text)

    // Divide the original string in words and calculate the CRC for each one
    text_split := strings.Split(text, " ")
    text_word_pos, text_word_crc := get_words_crc_pos(text_split, available_words)

    //fmt.Println("Words Pos:", text_word_pos)
    //fmt.Println("Words CRC:", text_word_crc)
    smallest_str := text_split
    found := false

    // Get all the possible sub sets in order to find one of the that sums the
    // same than all the given words
    main_loog: for subset_lenght := len(available_words); subset_lenght <= len(text_word_crc); subset_lenght++ {
	//fmt.Println("Len:", subset_lenght)

	for pos := 0; pos <= len(text_word_crc) - subset_lenght; pos++ {
		text_len := text_word_pos[pos + subset_lenght - 1] + 1 - text_word_pos[pos]

	    if text_len < len(smallest_str) {
		subset := text_word_crc[pos:pos + subset_lenght]
		//fmt.Println("SubSetPos:", text_word_pos[pos:pos + subset_lenght])

		if Sum(&subset) == words_total_weight {
		    found = true
		    smallest_str = text_split[text_word_pos[pos] : text_word_pos[pos + subset_lenght - 1] + 1]
		    //fmt.Println("NewSmalles:", text_len, smallest_str)

		    // We have a winner :), a smallest string can't be found
		    if len(smallest_str) == len(available_words) {
			break main_loog
		    }
		}
	    }
	}
    }

    if !found {
        fmt.Println("NO SUBSEGMENT FOUND")
    } else {
        fmt.Println(strings.Join(smallest_str, " "))
    }
}
